core/test/mpi/base/bindings.cpp

```diff
@@ -0,0 +1,758 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+
+
+#include "core/test/utils.hpp"
+
+
+template <typename T>
+class MpiBindings : public ::testing::Test {
+protected:
+    using value_type = T;
+    MpiBindings() : ref(gko::ReferenceExecutor::create()) {}
+
+    std::shared_ptr<gko::Executor> ref;
+
+    void assert_equal_vectors(std::vector<value_type>& vec_1,
+                              std::vector<value_type>& vec_2)
+    {
+        ASSERT_EQ(vec_1.size(), vec_2.size());
+        for (auto i = 0; i < vec_1.size(); ++i) {
+            EXPECT_EQ(vec_1[i], vec_2[i]);
+        }
+    }
+
+    void assert_equal_arrays(gko::Array<value_type>& array_1,
+                             gko::Array<value_type>& array_2)
+    {
+        ASSERT_EQ(array_1.get_num_elems(), array_2.get_num_elems());
+        for (gko::size_type i = 0; i < array_1.get_num_elems(); ++i) {
+            EXPECT_EQ(array_1.get_const_data()[i], array_2.get_const_data()[i]);
+        }
+    }
+};
+
+TYPED_TEST_SUITE(MpiBindings, gko::test::PODTypes, TypenameNameGenerator);
+
+
+TYPED_TEST(MpiBindings, CanSetADefaultWindow)
+{
+    gko::mpi::window<TypeParam> win;
+    ASSERT_EQ(win.get(), MPI_WIN_NULL);
+}
+
+
+TYPED_TEST(MpiBindings, CanCreateWindow)
+{
+    auto data = std::vector<TypeParam>{1, 2, 3, 4};
+    auto comm = gko::mpi::communicator::create_world();
+    auto win =
+        gko::mpi::window<TypeParam>(data.data(), 4 * sizeof(TypeParam), comm);
+    ASSERT_NE(win.get(), MPI_WIN_NULL);
+    win.lock_all();
+    win.unlock_all();
+}
+
+
+TYPED_TEST(MpiBindings, CanSendAndRecvValues)
+{
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto recv_array = gko::Array<TypeParam>{this->ref};
+    if (my_rank == 0) {
+        auto send_array = std::vector<TypeParam>{1, 2, 3, 4};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<TypeParam>(send_array.data(), 4, rank, 40 + rank,
```

Comment:
Same for all function calls throughout this file: This could be inferred from the parameter types.
```suggestion
                gko::mpi::send(send_array.data(), 4, rank, 40 + rank,
```